theHarvester/discovery/dnssearch.py

```diff
@@ -40,3 +59,191 @@ def process(self):
         return results
+
+#####################################################################
+# DNS REVERSE
+#####################################################################
+
+
+IP_REGEX = r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}'
+PORT_REGEX = r'\d{1,5}'
+NETMASK_REGEX = r'\d{1,2}|' + IP_REGEX
+NETWORK_REGEX = r'\b({})(?:\:({}))?(?:\/({}))?\b'.format(
+    IP_REGEX,
+    PORT_REGEX,
+    NETMASK_REGEX)
+
+
+def serialize_ip_range(
+        ip: str,
+        netmask: str = '24') -> str:
+    """
+    Serialize a network range in a constant format, 'x.x.x.x/y'.
+
+    Parameters
+    ----------
+    ip: str.
+        A serialized ip in the format 'x.x.x.x'.
+        Extra information like port (':z') or subnet ('/n')
+        will be ignored.
+    netmask: str.
+        The subnet subdivision, represented by a 2 digit netmask.
+
+    Returns
+    -------
+    out: str.
+        The network OSI address, like '192.168.0.0/24'.
+    """
+    __ip_matches = re.search(NETWORK_REGEX, ip, re.IGNORECASE)
+    __ip = __ip_matches.group(1)
+    __netmask = netmask if netmask else __ip_matches.group(3)
+    if __ip and __netmask:
+        return str(IPv4Network('{}/{}'.format(__ip, __netmask), strict=False))
+    elif __ip:
+        return str(IPv4Network('{}/{}'.format(__ip, '24'), strict=False))
+    else:   # invalid input ip
+        return ''
+
+
+def list_ips_in_network_range(
+        iprange: str) -> List[str]:
+    """
+    List all the IPs in the range.
+
+    Parameters
+    ----------
+    iprange: str.
+        A serialized ip range, like '1.2.3.0/24'.
+        The last digit can be set to anything, it will be ignored.
+
+    Returns
+    -------
+    out: list.
+        The list of IPs in the range.
+    """
+    try:
+        __network = IPv4Network(iprange, strict=False)
+        return [__address.exploded for __address in __network.hosts()]
+    except Exception:
+        return []
+
+
+async def reverse_single_ip(
+        ip: str,
+        resolver: DNSResolver) -> Awaitable[str]:
+    """
+    Reverse a single IP and output the linked CNAME, if it exists.
+
+    Parameters
+    ----------
+    ip: str.
+        The IP to reverse.
+
+    Returns
+    -------
+    out: str.
+        The corresponding CNAME or None.
+    """
+    try:
+        __host = await resolver.gethostbyaddr(ip)
+        return __host.name if __host else ''
+    except Exception:
+        return ''
+
+
+async def reverse_all_ips_in_range(
+        iprange: str,
+        callback: Callable) -> None:
+    """
+    Reverse all the IPs stored in a network range.
+    All the queries are made concurrently.
+
+    Parameters
+    ----------
+    iprange: str.
+        An IPv4 range formated as 'x.x.x.x/y'.
+        The last 2 digits of the ip can be set to anything,
+        they will be ignored.
+    callback: Callable.
+        Arbitrary postprocessing function.
+
+    Returns
+    -------
+    out: None.
+    """
+    __resolver = DNSResolver(timeout=4)
```

Comment:
While setting up the DNSResolver the user may also specifiy the dns server to make the requests for example with dns.resolver you can do [this](https://github.com/L1ghtn1ng/theHarvester/blob/dev/theHarvester/discovery/dnssearch.py#L13). With aiodns you can pass in a list of nameservers when instantiate the DNSResolver see [here](https://github.com/saghul/aiodns/blob/master/aiodns/__init__.py#L46) for more information. Please add that functionality to allow the user to pass a dns server into this function to make the requests.

Reply:
ok, should I add a CLI argument for the user?

Reply:
This is already a CLI argument see [here](https://github.com/laramies/theHarvester/blob/master/theHarvester/__main__.py#L31)

Reply:
oh wow I totally looked at this and missed it!..
ok this should do it.